kie-server-parent/kie-server-services/kie-server-services-task-assigning/kie-server-services-task-assigning-planning/src/main/java/org/kie/server/services/taskassigning/planning/SolverHandlerContext.java

```diff
@@ -49,11 +67,113 @@ public void clearProcessedChangeSet() {
         lastProcessedChangeSetId = -1;
     }
 
-    public LocalDateTime getLastModificationDate() {
-        return lastModificationDate;
+    /**
+     * Registers the time of the last change processed for a task.
+     * @param taskId identifier of the task to register.
+     * @param changeTime the task change time to register.
+     */
+    public void setTaskChangeTime(long taskId, LocalDateTime changeTime) {
+        taskChangeTimes.put(taskId, changeTime);
+    }
+
+    /**
+     * Indicates if a change has already been processed for a given task.
+     * @param taskId identifier of the task to query.
+     * @param changeTime the task change time to query.
+     * @return true if the change has already been processed, false in any other case.
+     */
+    public boolean isProcessedTaskChange(long taskId, LocalDateTime changeTime) {
+        return changeTime.equals(taskChangeTimes.get(taskId));
+    }
+
+    /**
+     * Removes all the registered task change times that occurred strictly before a given time.
+     * @param untilLocalDateTime the time for filtering the changes to remove.
+     */
+    public void clearTaskChangeTimes(LocalDateTime untilLocalDateTime) {
+        if (untilLocalDateTime != null) {
+            final List<Long> removableChanges = taskChangeTimes.entrySet().stream()
+                    .filter(entry -> untilLocalDateTime.compareTo(entry.getValue()) > 0)
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toList());
+            removableChanges.forEach(removableChange -> taskChangeTimes.remove(removableChange));
```

Comment:
I think the `forEach` can be called directly from the stream, right after the `map` call.

Reply:
the forEach can be chained after the .collect(Collectos.toList()) but not not after the map, otherwise a java.util.ConcurrentModificationException will be produced